### PR TITLE
feat/stability-rate

### PR DESCRIPTION
**Pull request by panukettu**
**General**

* `UniswapV2Oracle.sol` - TWAP capable, per-pair configurable Uniswap V2 oracle
* `LibStabilityRate.sol` - Stability rate related state update operations and getters
* `StabilityRateFacet.sol` - Facet for initializing, configuring and getting stability rate related state variables
* `InterestRateState.sol` - Diamond-storage for Interest Rate (only stability rate so far) related storage and layout

**Implementation**

The AMM oracle is deployed and then configured in the protocol by calling `setAmmOracle` in `ConfigurationFacet`.

Main integration between AMM oracle is found in `calculateStabilityRate` which calls `getPriceRate` that gets the current price rate of the AMM price against the oracle price. 

Flow of the stability rate:
1. Price rate is queried,  0.95 RAY = -5% premium, 1.05 RAY = +5% premium, 1 = Exact price match
2. Stability rate is calculated according to whether the price is within the delta range of exact price.
3. A cumulative debt index is calculated that updates on each `mintKreskoAsset`, `repayKreskoAsset` or a call to `updateStabilityRateAndIndexForAsset`   

This PR separates `ActionFacet` to separate `KreskoAssetFacet` and `CollateralAssetFacet`. 

Changes to debt-accounting are implemented in `LibMint` and `LibRepay` which converts the amount supplied into a debt-index adjusted amount when calling `mintKreskoAsset` and `repayKreskoAsset`.

**Misc**
* `WadRay.sol` - Ray-math library (from Aave)
* `Percentages.sol` - Percentage-math library (from Aave)

**NOT FEATURE COMPLETE**

* Separate scope of work is required for the repayment of debt in full
* Due to the separation, some tests cases are marked as "pending" as they require this implementation
* Separate scope of work is required for validating the AMM liquidity the oracle queries.

### CODE REVIEW

File: src/contracts/minter/facets/StabilityRateFacet.sol [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+ * @author Kresko
+ * @notice Stability rate related views and state operations
+ * @dev Uses both MinterState (ms) and InterestRateState (irs)
+ */
+contract StabilityRateFacet is DiamondModifiers {
+    using WadRay for uint256;
+    using LibStabilityRate for StabilityRateConfig;
+
+    // Stability Rate setup struct
+    struct StabilityRateSetup {
```
[the remark this answers is not in the document]
Comment (diofinance) [reply src/contracts/minter/facets/StabilityRateFacet.sol:23]: We should also decide if we want to use a `TWAL` (time-weighted average liquidity) to determine at which size we'll start respecting an AMM pool's size and turn on the rate

**Review by diofinance — COMMENTED** [review]
Submitting what I've got for now and will expand on some comments during the revision process. The core protocol functionality looks very complete with the recent additional changes.

File: src/abi/interest/DebtToken.sol [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+import {LibWadRay} from "./LibWadRay.sol";
+import {DebtTokenBase} from "./DebtTokenBase.sol";
+import {LibInterestRate} from "./LibInterestRate.sol";
+import {AssetConfig, InterestRateMode} from "./DebtTokenTypes.sol";
+
+/**
+ * @title VariableDebtToken
+ * @notice Implements a variable debt token to track the borrowing positions of users
+ * at variable rate mode
+ * @author Aave
```
[the remark this answers is not in the document]
Comment (panukettu) [reply src/abi/interest/DebtToken.sol:13]: this is just a reference copy of aave variable debt token

[the remark this answers is not in the document]
Comment (panukettu) [reply src/abi/interest/DebtToken.sol:13]: we are not using this

File: src/contracts/libs/WadRay.sol [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+     * @param a Wad
+     * @param b Wad
+     * @return The result of a*b, in wad
+     **/
+    function wadMul(uint256 a, uint256 b) internal pure returns (uint256) {
+        if (a == 0 || b == 0) {
+            return 0;
+        }
+
+        require(a <= (type(uint256).max - halfWAD) / b, "Multiplication overflow");
```
[the remark this answers is not in the document]
Comment (panukettu) [reply src/contracts/libs/WadRay.sol]: this was V2 aave copy, updated to V3

File: src/contracts/kreskoasset/UniswapV2Oracle.sol [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        pairs[_pair].updatePeriod = _updatePeriod;
+    }
+
+    function update(address _pair) external {
+        (uint256 price0Cumulative, uint256 price1Cumulative, uint32 blockTimestamp) = currentCumulativePrices(_pair);
+
+        PairData storage data = pairs[_pair];
+        require(data.blockTimestampLast != 0, "!u:exists");
+
+        uint32 timeElapsed = blockTimestamp - data.blockTimestampLast; // overflow is desired
```
[the remark this answers is not in the document]
Comment (panukettu) [reply src/contracts/kreskoasset/UniswapV2Oracle.sol:129]: First require checks that pair exists and the second one is enforcing the TWAP. (updatePeriod)

**Review by diofinance — APPROVED** [review]
Revisions + new changes LGTM

**Pull request merged**